### ISSUE TITLE
Questions about being able to use webhooks and vmq_diversity (Lua script) auth in parallel

### ISSUE THREAD

**Issue by puresilk** [vernemq/vernemq#1199]
### VerneMQ basics
System & VerneMQ version like in my other issue #1163
```
vernemq version
```
shows an empty line for me. I believe I use version 1.7.1, because of the following line in my Dockerfile:
```
ARG VERNEMQ_GIT_REF=1.7.1
```

![image](https://user-images.githubusercontent.com/4606111/58440168-89850c00-80d8-11e9-81fa-014f5c5ca930.png)

#### possibly relevant entries in vernemq.conf.local

This is an excerpt:
```
allow_anonymous = off
listener.max_connections = infinity
listener.ws.jsclient = 0.0.0.0:8080
listener.ws.piclient = 0.0.0.0:8078
listener.ws.jsclient.mountpoint = jsclient
listener.ws.piclient.mountpoint = piclient
plugins.vmq_passwd = off
plugins.vmq_acl = off
plugins.vmq_diversity = on
plugins.vmq_webhooks = on
vmq_diversity.script_dir = ./share/lua
vmq_diversity.auth_mongodb.enabled = on
(... and other relevant settings for mongodb which worked when webhooks had been disabled)
log.console = both
log.console.level = debug
nodename = VerneMQ@127.0.0.1
vmq_webhooks.js_auth_on_register.hook = auth_on_register
vmq_webhooks.js_auth_on_publish.hook = auth_on_publish
vmq_webhooks.js_auth_on_register.endpoint = http://pcp_picockpit:3000/mqtt/js-auth-on-register
vmq_webhooks.js_auth_on_publish.endpoint = http://pcp_picockpit:3000/mqtt/js-auth-on-publish

```

### The basic idea / setup
I want to have two "classes" of clients which connect on different ports / listeners. 
Each class of client should have it's own authentication and authorization. 

I decided, due to the way my application logic is implemented, to use webhooks and mongodb (Lua scripts) for the separate client classes.

As currently VerneMQ does not support a separated auth setup per listener (i.e. enabling certain plugins / webhooks for a certain listener), my next idea was to set a mountpoint on the individual port, and check in the plugins (webhooks and lua) if I should handle it, otherwise to let the request pass. 

I understand from the documentation that VerneMQ will pass in the auth requests to each app in return - unless one of them returns an error. 

I believe that if auth_commons.lua returns "false" in this part:
```
function auth_on_publish(pub)
    print("***auth_on_publish debug*** (LUA)")
    return false
end
```
That VerneMQ will pass control to the webhooks, and call the webhook. However that is not what happens.

As per my current understanding, Lua return values differ from values if the plugin would be implemented in Erlang - therefore a "false" which is returned by Lua, will be (ultimately?) treated as if a "next" had been returned. Will be happy to be corrected if this is a misunderstanding from my side.

### The webhooks

```
root@jalpine_morpheus:/vernemq# vmq-admin webhooks show
+----------------+--------------------------------------------------+-------------+
|      hook      |                     endpoint                     |base64payload|
+----------------+--------------------------------------------------+-------------+
|auth_on_register|http://pcp_picockpit:3000/mqtt/js-auth-on-register|    true     |
|auth_on_publish |http://pcp_picockpit:3000/mqtt/js-auth-on-publish |    true     |
+----------------+--------------------------------------------------+-------------+
```
I have also tried adding a auth_on_subscribe hook, but this did not change things. Do I need to add on_publish, etc. hooks, if I am not interested in them?

On the other side of the webhooks I am using the following Crystal code:
```
	post "/mqtt/js-auth-on-register" do |env|
		puts("entering js-auth-on-register")
		if env.request.headers.has_key?("vernemq-hook")
			puts(env.request.headers["vernemq-hook"])
			env.response.print "{\"result\": \"ok\"}" # also possible: result: next; result: {"error": "not_allowed"}
			env.response.close
		end
	end

	post "/mqtt/js-auth-on-publish" do |env|
		puts("entering js-auth-on-publish")
		if env.request.headers.has_key?("vernemq-hook")
			puts(env.request.headers["vernemq-hook"])
			env.response.print "{\"result\": \"ok\"}" # also possible: result: next; result: {"error": "not_allowed"}
			env.response.close
		end
	end

	post "/mqtt/js-auth-on-subscribe" do |env|
		puts("entering js-auth-on-subscribe")
		if env.request.headers.has_key?("vernemq-hook")
			puts(env.request.headers["vernemq-hook"])
			env.response.print "{\"result\": \"ok\"}" # also possible: result: next; result: {"error": "not_allowed"}
			env.response.close
		end
	end
```
This code should allow everything, as per my current understanding. 

### Expected behavior
The webhook for auth_on_subscribe is called on client connection.
The webhook for auth_on_publish is called
client can publish message.

### Actual behaviour

The webhook for auth_on_subscribe **is** called, I see this in my server.
```
2019-05-27 20:55:11 UTC 200 POST /mqtt/js-auth-on-register 256.0µs
entering js-auth-on-register
auth_on_register
```

The webhook for auth_on_publish **is not** called. Only the Lua script is called. 
The client (Javascript, Paho, connecting using websockets) can not publish the message, and is disconnected. 

```
MQTT connection lost (8): AMQJS0008I Socket closed.
```

```
jalpine_morpheus | ***auth_on_publish debug*** (LUA)
jalpine_morpheus | 21:09:27.420 [error] can't auth publish [<<"my_user">>,{"jsclient",<<"my_clientid">>},0,[<<"test">>],<<"Hello world!">>,false] due to not_authorized
jalpine_morpheus | 21:09:27.420 [debug] ws session normally stopped
```
![image](https://user-images.githubusercontent.com/4606111/58439930-39f21080-80d7-11e9-81ce-bab04d1c9f16.png)

Additionally, I believe that now the client which is supposed to be authenticated through the MongoDB Lua script is failing to connect to VerneMQ - I have however not debugged this in depth. I believe if the other issue is resolved / understood by me, that I will be able to debug these client connection problems easier, as well.   

P.S: Thanks for the information about the mountpoints #533  @larshesel , it helped me during setup !

### Other approaches

I'm also open to suggestions for other approaches to set up my auth architecture. One approach I considered would be to do everything from the Lua scripts. I'm also worried about the performance impact if the webhook is called for the other client class, which is only supposed to be authenticated via Lua / MongoDB. Is there a way to stop the webhook of trying to authenticate users on a certain mountpoint / listener currently? 

Possibly also two bridged VerneMQ instances might be a solution?

**Comment by puresilk** [comment]
I have managed to solve the problem, and am analyzing it in detail in this blog post.

https://pi3g.com/2019/06/04/debugging-vernemq-connection-authentication-problems-for-dual-authentication-with-webhooks-and-vmq_diversity/


TL;DR the solution is **to not return anything from the Lua script**, in case you do not want to handle the message. As soon as you return "false", the auth request is finalized, and the client is rejected.

![image](https://user-images.githubusercontent.com/4606111/58916124-f6288800-8722-11e9-96fe-c93dfec892c6.png)

**Note to the VerneMQ dev team: this is IMHO a stumbling block to first-time users, I feel a hint in the documentation as to the return values might be a time saver for people dealing with the Lua scripts.**

For anyone else struggling with Lua scripting for VerneMQ (return values specifically) I also recommend to study my other extensive blog post:

https://pi3g.com/2019/05/12/understanding-erlang-lua-luerl-for-vernemq-mongodb-auth_on_register-hook/
